### ISSUE TITLE
Due to accumulation of sequence, validation is so slow.

### ISSUE THREAD

**Issue by kaiyikang** [speechbrain/speechbrain#986]
I followed the recipe `LibriSpeech/ASR/seq2seq` and replaced the encoder with my custom model. The configuration and train.py haven't changed much to prevent bugs.

Because the validation process is very slow, I print the `wavs.shape`, `feats.shape`, `x.shape`. It looks like this: 
~~~python
    def compute_forward(self, batch, stage):
        batch = batch.to(self.device)
        wavs, wav_lens = batch.sig
        tokens_bos, _ = batch.tokens_bos
        wavs, wav_lens = wavs.to(self.device), wav_lens.to(self.device)
        # Add augmentation if specified
        if  stage == sb.Stage.TRAIN:
            if hasattr(self.modules, "env_corrupt"):
                wavs_noise = self.modules.env_corrupt(wavs, wav_lens)
                wavs = torch.cat([wavs, wavs_noise], dim=0)
                wav_lens = torch.cat([wav_lens, wav_lens])
                tokens_bos = torch.cat([tokens_bos, tokens_bos], dim=0)
            if hasattr(self.hparams, "augmentation"):
                wavs = self.hparams.augmentation(wavs, wav_lens)

        # Forward pass
        print("wavs.shape:",wavs.shape) # DEBUG
        feats = self.hparams.compute_features(wavs)
        feats = self.modules.normalize(feats, wav_lens)
        print("feats.shape:",feats.shape) # DEBUG
        x = self.modules.enc(feats.detach())
        print("x.shape:",x.shape) # DEBUG
        e_in = self.modules.emb(tokens_bos)  # y_in bos + tokens
        h, _ = self.modules.dec(e_in, x, wav_lens)

        # Output layer for seq2seq log-probabilities
        logits = self.modules.seq_lin(h)
        p_seq = self.hparams.log_softmax(logits)

        # Compute outputs
        if stage == sb.Stage.TRAIN:
            current_epoch = self.hparams.epoch_counter.current
            if current_epoch <= self.hparams.number_of_ctc_epochs:
                # Output layer for ctc log-probabilities
                logits = self.modules.ctc_lin(x)
                p_ctc = self.hparams.log_softmax(logits)
                return p_ctc, p_seq, wav_lens
            else:
                return p_seq, wav_lens
        else:
            if stage == sb.Stage.VALID:
                p_tokens, scores = self.hparams.valid_search(x, wav_lens)
            else:
                p_tokens, scores = self.hparams.test_search(x, wav_lens)
            return p_seq, wav_lens, p_tokens
~~~
The shapes of variables in the training phase and validation phase are very different:
**Training**:
![image](https://user-images.githubusercontent.com/22473900/132988974-4df7c6a0-0849-4765-8983-7a18d992bd0d.png)
**Validation**:
![image](https://user-images.githubusercontent.com/22473900/132988721-72ec66c9-ca1b-4b07-bd74-16f4da526db4.png)

Due to the accumulation of the length of input `wavs`, validation is getting slower and slower.

Is the accumulation in validation correct? Or maybe my custom model has a problem with generating too-long sequences?  (p.s. run with 2 GPUs and added --data_parallel_backend)

Thanks so much.

About more details, you can see the train.py and .yaml files here: [google drive link](https://drive.google.com/drive/folders/1q4obmR2CJq6dofVkUsuWzcT8jl9YCqlV?usp=sharing)

**Comment by mravanelli** [comment]
Hi,
the reason why it is getting slower and slower is because the system sorts
sentences in ascending order according to their length. This way, shorter
sentences are processed first, and longer ones later (there is no
accumulation of the length input, we just process progressively longer
sentences). That's expected. If you would like to speed up validation, you
can reduce the batch size for the valid search.

**Comment by kaiyikang** [comment]
Thanks for your quick response. (always :) )
Yes, after some small tries, it actually depends on the dev set.
The increased length confuses me.
